other/mentoring-task/design_patterns_task_1.rb

```diff
@@ -0,0 +1,81 @@
+# Script with objective of provide design patterns examples (3 design patterns)
+
+"--------- SINGLETON -----------"
+
+class Connector
+  private_class_method :new
+
+  @@instance = new
+
+  def self.get_connection
+    return @@instance
+  end
+end
+
+connector_one = Connector.get_connection
+connector_two = Connector.get_connection
+
+# Bot references are using same instance
+puts "#{connector_one.object_id}"
+puts "#{connector_two.object_id}"
+
+"--------- NULL OBJECT -----------"
+
+class NullCar
+  def drive
+    puts "No car assigned"
+  end
+end
+
+class Car
+
+  def drive
+    puts "driving the car"
+  end
+
+end
+
+class User
+  @car = nil
+
+  def initialize(car)
+    @car = car
+  end
+
+  def get_car
+    @car || NullCar.new
+  end
+
+end
+
+first_user = User.new(nil)
+second_user = User.new(Car.new)
+
+first_user.get_car.drive
+second_user.get_car.drive
+
+"--------- DECORATOR -----------"
+
+class House 
+  def description
+    puts "This house have windows"
+    puts "This house have main door"
+  end
+end
+
+class HouseWithFirePlace
+  @house  
+
+  def initialize(house)
+    @house = house
+  end
+
+  def description
+    puts "This house have a fire place"
+    @house.description
+  end

```

Comment:
In well-established Ruby projects we tent to apply private `attr_reader` like this:
```suggestion
  def description
    puts "This house have a fire place"
    house.description
  end
  
  private
  
  attr_reader :house
```
Using instance variables directly (`@something`) can be considered as a bad practice. First of all, because use undefined instance variable will return `nil`, and it's so easy to misprint the name, so you won't catch the bug on the runtime.
Besides, we don't use `attr_accessor` as it provides a wrapper to mutate instance variable, while good practice is to write ruby classes as kind of pure functions, with no side-effects. So, we use exactly `attr_reader`.